rfc/cart-api-v1.md

```diff
@@ -0,0 +1,251 @@
+
+* The endpoint will also respond **HTTP 404** if the corresponding cart is not found.
+
+### Other APIs
+
+#### `POST /resources/carts/{cart_id}/items/query`
+
+Save the cart items in batch by query.
+
+| Param Type | Param Name | Association |
+| --- | --- | --- |
+| URL Path | `cart_id` | `Cart.id` |
+
+TL;DR:
+* Respond **HTTP 202** on successfully accepting a request.
+* Use `POST /resources/cart-item-batch-jobs/{job_id}` to check the async job status.
+
+The response body for HTTP 202 will be `CartItemBatchJob`.
+
+##### Technical Considerations
+
+As this operation may require more than 30 seconds to run, the batch operation is divided into three major parts: processing requests, job updates, and job reports.
+
+> See [the overall diagram describing the process](https://github.com/DataBiosphere/azul/raw/rfc/cart-api/rfc/DNAstack%20-%20General%20-%20Batch%20Cart%20Item%20Creation%20API%20Diagram.png).
+
+###### Processing requests
+
+> See "Creation Request" in the diagram.
+
+When the endpoint `λ1` receives a request with filters, the endpoint will delegate the query to `λ2` via **the request queue**. At this point, the client will not be able to figure out whether the new cart items based on the query are in the cart. The client will have to track the progress by sending request to `λ5` for an update.
+
+> This is to offload querying ElasticSearch and publishing message to **the job part queue** to `λ2` as publishing messages seems to take a long time to finish.
+
+> Also, either `λ1` or `λ2` will first the total number of hits so that the status check part can tell whether or not the batch job is done.
+
+When `λ2` receives a message with a requested filter, it will run the query, iterate the result and send a list of ready-to-write `CartItem` to `λ3` (in multiple batches) via **the job part queue**.
+
+> With [the size limit on SQS](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-limits.html), the message publication will be more frequent than ES scanning.
+
+Then, for each batch, `λ3` will put each ready-to-write `CartItem` to DynamoDB. At the end of each batch, `λ3` will notify `λ4` that it finished writing the given batch via **the job part complete notification queue**.
+
+###### job Updates
+
+For each notice in **the job part complete notification queue**, `λ4` will either create a new `CartItemBatchJob` or try to update the corresponding one with the updated number of remaining parts.
+
+> **The job part complete notification queue** is a FIFO queue to minimize race condition.
+
+> The approach to avoid race conditions is based on [the doc on optimistic locking for DynamoDB with Java](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/DynamoDBMapper.OptimisticLocking.html) and [the use of `ConditionExpression` in `DynamoDBResource.put_item`](https://github.com/boto/boto3/issues/1550) where `CartItemBatchJob.remaining_parts` is acting as version.
```

Comment:
I think this is a more relevant link https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/WorkingWithItems.html#WorkingWithItems.ConditionalUpdate

The java optimistic locking isn't really relevant to boto3